### ISSUE TITLE
Is it usable?

### ISSUE THREAD

**Issue by Xerus2000** [sghpjuikit/player#1]
Is it possible to run it somehow? Are all these features already implemented or plans?

**Comment by sghpjuikit** [comment]
Hi, thank you for showing interest.

Yes it is possible to run the application. From IDE or running it as a jar or as as a standalone application with JRE included, or as an exe that invokes the jar. I have done all of these. I've been using the application as a Winamp substitute for 3-4 years already and it works (except for minor UX issues and bugs here and there. 

The only problem with usage may be somewhat untraditiona/unfamiliar UI that is not made for an average user in mind. Plus some minor hickups that average user may find important (depends on use case).

All the mentioned features are implemented.

Currently, I'm turning adoptin Kotlin and as such making huch changes to the code, fixing numerous bugs and adding numerous improvements to the existing features. Unfortunately, due to negligence, I've just lost like 3 months of work in my local repository.

Anyway: With java 9 now finally released and Kotlin being production ready, I'm about to make a release. ETA is 2 months (I'm about to entirely focus on this instead of daily job for a while).
What this means:
- In 2 weeks: Simplified loading of the project in an IDE for potential developers, JDK9 support, etc
- In 2 months: Runnable application (jar +  standalone + exe) with Windows fully supported and OSX/Unix untested.

If you are interested in using the application, I kindly ask you to be patient for just a little while.
Uf you are interested in developing or building the project, then I'll shortly update the project to a point where you'll should easily be able to do that. One of the hurdles is that the project is not mavenized.

**Comment by Xerus2000** [comment]
I am interested in developing a music player with JavaFX with some special features, but seeing your project, I might as well collaborate, as to not having to design everything from scratch.

What I am interested in is a replacement for foobar2000, that means that it needs to be customizable (fxml?) and should provide some kind of plugin-system.
My own concept, why I've wanted to do this in the first place, is that I want a self-learning music player. I already have my own concept partly implemented, but with a very poor interface.
Compatibility with JavaFXPorts is also a goal of mine, as to create a cross-platform application. Many android players I found have been lacking correct art loading, artist and genre tag separation etc.

I've never really used maven, only to fetch dependencies for gradle. For what I've used it so far, gradle definitely did its job.

To summarise: Yes, I would like to contribute, and if we come to common terms, not only on small stuff!

**Comment by sghpjuikit** [comment]
I am not sure you will be able to bend this project to your will, as it is already quire big (thus requiring quite a dedication). Of course it is possible to take out only parts, but the separation of modules is not ideal and there is also bunch of static classes in the way as is entire bunch of util classes that very well may be a separate project (if the dependencies were properly taken care of).

I understand that starting from scratch is not the best idea, but sometimes it is the only one. Reusing other people's work is important, but not so easy when you want to heavily modify or discard large parts.

That being said

- I'm open to collaboration (let me set things up a bit on my end for this), small or big. 

- What kind of customization are we talking about? In my experience fxml files are not terribly customizable and I cant imagine user opening SceneBuilder and making his own user interface. Wiring fxml with behavior (code) is also necessary. Alternatively, it is possible to write custom DSL to define ui declaratively. Css is also pretty difficult to get right. Altogether, I think this is something extremely hard.

- As for the customization this project already supports: functional UI modules. Basically you write a class implementing an interface and the application will discover and allow user to use it in UI in a predefined way. The application defines the way this UI modules can be displayed (tabs, freeform, splits, popups). These modules are standard java classes, which can load fxml and use custom css if necessary. The application has no dependency on these modules, so in theory two users could have completely different experience and use for the application. These modules provide almost all functionality (from the user's perspective) so there is immense room for taking the project in direction you wish (without influencing other users and overal idea).

- Without knowledge of the Kotlin language you will have a hard time to get anywhere, as I'm porting substantial parts of the project into Kotlin. However it is easy, much better than Java and interops with it nicely, so I recommend checking it out.

**Comment by Xerus2000** [comment]
Kotlin looks useful. As a programmer, expanding ones horizon is always a good thing ;)

As to customization... This is how my foobar2000 currently looks:
![image](https://user-images.githubusercontent.com/13354331/32150702-9cdba9ee-bd16-11e7-8a05-0ce666f26eb0.png)
On the starting page you wrote about a modular user interface, which was what i was talking about, with custom components like the waveform-seekbar. Although it is not the most important thing, it does come in handy.

The ideas I have could fit on top of any player I guess, but not having access to the code of the player itself limits the ways as to how to do that. Basically, my Player is supposed to collect data while I listen and then form connections between Songs and ratings for the Songs. Then, while shuffling, it will use these to intelligently find the next Song. Alternatively the user can define a "mood" (or playlist) with a few Songs selected and the player will then find fitting Songs for it. This way I want to make my 1500 Songs collection actually usable without having to manually search and categorise forever...

A few questions: 
Are you on Discord?
Does your player already support ReplayGain?
Are you familiar with Smart Playlists, a feature of foobar?

**Comment by sghpjuikit** [comment]
I believe most of what you would like to do could be implemented fairly straightforwardly here, on top of the modular UI mentioned. It is possible to create a waveform component or control and integrate it into existing playback controling component or create new one. It would also be possible to collect the data you are interested in, create song orders based on the data and so on. So yeah, looks very promising.

- Yes I am on Discord, we can get in touch tomorrow at night.
- No it does not, I havent played around with this feature. I also use 2 player implementations based on the format of the song playing, so implementing something like this would take extra effort. I was thinking about using vlcj in the future, which I guess should handle this kind of thing.
- No I am not, but I think playlist features are simple to implement with proper design.

**Comment by Xerus2000** [comment]
- Add me via Xerus#8428
- For the implementation of ReplayGain the format does not matter, at least if you don't scan it yourself. Since all my songs are replaygain-scanned by foobar, I already implemented this in my mockup-player by reading the Tag and adjusting the volume accordingly. Scanning would be ideal in the long term, but not immediately necessary.
- after creating such a great project, I don't believe that adding VLC as dependency would be a good idea.
- What about crossfading? I nearly consider it a necessity for a smooth listening experience
- I've seen that you can create versatile worded queries. These could be directly used as autoplaylists (a playlist that contains all the Songs that match the query and updates regularly)!
